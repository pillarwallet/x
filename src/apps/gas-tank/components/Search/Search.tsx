/* eslint-disable @typescript-eslint/no-non-null-asserted-optional-chain */
import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  useRef,
  useState,
} from 'react';
import { TailSpin } from 'react-loader-spinner';
import { useLocation, useNavigate } from 'react-router-dom';
import { isAddress } from 'viem';
import {
  Token,
  chainNameToChainIdTokensData,
} from '../../../../services/tokensData';
import { PortfolioData } from '../../../../types/api';
import {
  formatExponentialSmallNumber,
  limitDigitsNumber,
} from '../../../../utils/number';
import SearchIcon from '../../assets/seach-icon.svg';
import { useTokenSearch } from '../../hooks/useTokenSearch';
import { SearchType, SelectedToken } from '../../types/tokens';
import { MobulaChainNames } from '../../utils/constants';
import { Asset, parseSearchData } from '../../utils/parseSearchData';
import Close from '../Misc/Close';
import Esc from '../Misc/Esc';
import Refresh from '../Misc/Refresh';
import PortfolioTokenList from './PortfolioTokenList';

interface SearchProps {
  setSearching: Dispatch<SetStateAction<boolean>>;
  isBuy: boolean;
  setBuyToken: Dispatch<SetStateAction<SelectedToken | null>>;
  setSellToken: Dispatch<SetStateAction<SelectedToken | null>>;
  chains: MobulaChainNames;
  setChains: Dispatch<SetStateAction<MobulaChainNames>>;
  walletPortfolioData?: PortfolioData;
  walletPortfolioLoading?: boolean;
  walletPortfolioFetching?: boolean;
  walletPortfolioError?: boolean;
  refetchWalletPortfolio?: () => void;
}


type SortKey = 'symbol' | 'price' | 'balance' | 'pnl';
type SortOrder = 'asc' | 'desc';

export default function Search({
  setSearching,
  isBuy,
  setBuyToken,
  setSellToken,
  chains,
  walletPortfolioData,
  walletPortfolioLoading,
  walletPortfolioFetching,
  walletPortfolioError,
  refetchWalletPortfolio,
}: SearchProps) {
  const { searchText, setSearchText, searchData, isFetching } = useTokenSearch({
    isBuy,
    chains,
  });
  const [searchType, setSearchType] = useState<SearchType>();
  const [sortKey, setSortKey] = useState<SortKey | null>(null);
  const [sortOrder, setSortOrder] = useState<SortOrder>('desc');

  const inputRef = useRef<HTMLInputElement>(null);
  const searchModalRef = useRef<HTMLDivElement>(null);

  const useQuery = () => {
    const { search } = useLocation();
    return new URLSearchParams(search);
  };

  const query = useQuery();

  const navigate = useNavigate();
  const location = useLocation();

  const removeQueryParams = () => {
    navigate(location.pathname, { replace: true });
  };

  useEffect(() => {
    inputRef.current?.focus();
    const tokenAddress = query.get('asset');

    // Only read asset parameter when in buy mode to prevent token address
    // from token-atlas showing in sell search
    if (isAddress(tokenAddress || '')) {
      setSearchText(tokenAddress!);
    }

    setSearchType(SearchType.MyHoldings);
  }, [query, setSearchText]);

  const handleClose = () => {
    setSearchText('');
    // It resets search type to MyHoldings if on sell screen
    setSearchType(SearchType.MyHoldings);
    setSearching(false);
    removeQueryParams();
  };

  // Click outside to close functionality
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        searchModalRef.current &&
        !searchModalRef.current.contains(event.target as Node)
      ) {
        handleClose();
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // ESC key to close functionality
  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        handleClose();
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleSort = (key: SortKey) => {
    if (sortKey === key) {
      setSortOrder(sortOrder === 'asc' ? 'desc' : 'asc');
    } else {
      setSortKey(key);
      setSortOrder('desc');
    }
  };

  const getSortIcon = (key: SortKey) => {
    if (sortKey !== key) return 'â†•';
    return sortOrder === 'asc' ? 'â†‘' : 'â†“';
  };

  const handleTokenSelect = (item: Asset | Token) => {
    if (isBuy) {
      // Asset type
      if ('chain' in item) {
        setBuyToken({
          name: item.name,
          symbol: item.symbol,
          logo: item.logo ?? '',
          usdValue: formatExponentialSmallNumber(
            limitDigitsNumber(item.price || 0)
          ),
          dailyPriceChange: -0.02,
          chainId: chainNameToChainIdTokensData(item.chain),
          decimals: item.decimals,
          address: item.contract,
        });
      } else {
        // Token type
        setBuyToken({
          name: item.name,
          symbol: item.symbol,
          logo: item.logo ?? '',
          usdValue: formatExponentialSmallNumber(
            limitDigitsNumber(item.price || 0)
          ),
          dailyPriceChange: -0.02,
          chainId: chainNameToChainIdTokensData(item.blockchain),
          decimals: item.decimals,
          address: item.contract,
        });
      }
    } else {
      const sellTokenData = {
        name: item.name,
        symbol: item.symbol,
        logo: item.logo ?? '',
        usdValue: formatExponentialSmallNumber(
          limitDigitsNumber(item.price || 0)
        ),
        dailyPriceChange: -0.02,
        decimals: item.decimals,
        address: item.contract,
      };

      if ('chain' in item) {
        setSellToken({
          ...sellTokenData,
          chainId: chainNameToChainIdTokensData(item.chain),
        });
      } else {
        setSellToken({
          ...sellTokenData,
          chainId: chainNameToChainIdTokensData(item.blockchain),
        });
      }
    }
    setSearchText('');
    // This keeps MyHoldings filter active when on sell screen
    if (!isBuy) {
      setSearchType(SearchType.MyHoldings);
    }
    setSearching(false);
    removeQueryParams();
  };

  return (
    <div
      className="flex items-center justify-center min-h-screen bg-transparent"
      data-testid="pulse-search-view"
    >
      <div
        ref={searchModalRef}
        className="flex flex-col w-full max-w-[95vw] sm:max-w-[600px] md:max-w-[700px] lg:max-w-[800px] xl:max-w-[900px] max-h-[90vh] sm:max-h-[80vh] bg-[#2A2A2A] border border-white/[0.05] rounded-2xl shadow-[0px_2px_15px_0px_rgba(18,17,22,0.5)] overflow-hidden"
        data-testid="pulse-search-modal"
      >
        {/* Fixed Header Section */}
        <div className="flex-shrink-0 p-3 pb-0">
          <div className="flex w-full items-start">
            <div className="flex items-center justify-center flex-1 h-10 bg-[#121116] rounded-[10px] mr-2 border-2 border-[#1E1D24] min-w-0">
              <span className="ml-2.5 flex-shrink-0">
                <img src={SearchIcon} alt="search-icon" />
              </span>
              <input
                ref={inputRef}
                type="text"
                className="flex-1 w-full ml-4 font-normal text-xs text-gray-500 bg-transparent border-none outline-none"
                value={searchText}
                data-testid="pulse-search-input"
                placeholder="Search tokens..."
                onChange={(e) => {
                  setSearchText(e.target.value);
                }}
              />
              {(searchText.length > 0 && isFetching) ? (
                <div className="mr-2.5 flex-shrink-0">
                  <TailSpin color="#FFFFFF" height={20} width={20} />
                </div>
              ) : (
                <div className="flex-shrink-0">
                  <Close onClose={handleClose} />
                </div>
              )}
            </div>
            <div className="flex items-center gap-1.5">
              <div className="w-10 h-10 bg-black rounded-[10px] p-0.5 pb-1 pl-0.5 pr-0.5">
                <Refresh
                  isLoading={walletPortfolioFetching}
                  onClick={refetchWalletPortfolio}
                  disabled={!refetchWalletPortfolio || walletPortfolioFetching}
                />
              </div>
              <div className="w-10 h-10 bg-black rounded-[10px] p-0.5 pb-1 pl-0.5 pr-0.5 relative cursor-pointer">
                <Esc onClose={handleClose} />
              </div>
            </div>
          </div>

          {/* MyHoldings Header */}
          <div className="flex gap-1.5 mt-4 mb-2" data-testid="pulse-search-filter-buttons">
            <div className="flex items-center">
              <p className="text-[13px] font-normal text-white tracking-[-0.26px] px-3">
                ðŸ’°My Holdings
              </p>
            </div>
          </div>
        </div>

        {/* Scrollable Content Section */}
        <div className="flex-1 min-h-0 px-3 pb-3 overflow-hidden">
          <div className="h-full overflow-y-auto overflow-x-hidden" style={{ maxHeight: 'calc(90vh - 140px)' }}>
            <table className="w-full table-fixed">
              <thead className="sticky top-0 bg-[#2A2A2A] z-10 border-b border-gray-700">
                <tr>
                  <th
                    className="text-left text-gray-400 text-xs font-medium px-2 pb-3 pt-2 cursor-pointer hover:text-white transition-colors w-2/5"
                    onClick={() => handleSort('symbol')}
                  >
                    Token/Price {getSortIcon('symbol')}
                  </th>
                  <th
                    className="text-center text-gray-400 text-xs font-medium px-2 pb-3 pt-2 cursor-pointer hover:text-white transition-colors w-1/5"
                    onClick={() => handleSort('balance')}
                  >
                    Balance {getSortIcon('balance')}
                  </th>
                  <th
                    className="text-right text-gray-400 text-xs font-medium px-2 pb-3 pt-2 cursor-pointer hover:text-white transition-colors w-2/5"
                    onClick={() => handleSort('pnl')}
                  >
                    Unrealized PnL/% {getSortIcon('pnl')}
                  </th>
                </tr>
              </thead>
              <tbody>
                <PortfolioTokenList
                  walletPortfolioData={walletPortfolioData}
                  handleTokenSelect={handleTokenSelect}
                  isLoading={walletPortfolioLoading}
                  isError={walletPortfolioError}
                  searchText={searchText}
                  sortKey={sortKey}
                  sortOrder={sortOrder}
                />
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  );
}
