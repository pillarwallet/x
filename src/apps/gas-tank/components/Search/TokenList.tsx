import { useState } from 'react';
import { Asset } from '../../utils/parseSearchData';
import RandomAvatar from '../../../pillarx-app/components/RandomAvatar/RandomAvatar';
import { getLogoForChainId } from '../../../../utils/blockchain';
import { chainNameToChainIdTokensData } from '../../../../services/tokensData';
import { SearchType, SortType } from '../../types/tokens';
import { formatBigNumber } from '../../utils/number';
import TokenPrice from '../Price/TokenPrice';
import TokenPriceChange from '../Price/TokenPriceChange';

export interface TokenListProps {
  assets: Asset[];
  handleTokenSelect: (item: Asset) => void;
  searchType?: SearchType;
}

export default function TokenList(props: TokenListProps) {
  const { assets, handleTokenSelect, searchType } = props;

  const [sort, setSort] = useState<{
    mCap?: SortType;
    volume?: SortType;
    price?: SortType;
    priceChange24h?: SortType;
  }>({});

  const handleSortChange = (
    key: 'mCap' | 'volume' | 'price' | 'priceChange24h'
  ) => {
    const sortType =
      // eslint-disable-next-line no-nested-ternary
      sort[key] === SortType.Down
        ? SortType.Up
        : sort[key] === SortType.Up
          ? SortType.Down
          : SortType.Up;

    assets.sort((a, b) => {
      if (sortType === SortType.Up) {
        return (b[key] || 0) - (a[key] || 0);
      }
      return (a[key] || 0) - (b[key] || 0);
    });

    setSort({
      mCap: undefined,
      price: undefined,
      priceChange24h: undefined,
      volume: undefined,
      [key]: sortType,
    });
  };

  if (assets) {
    return (
      <>
        {assets.map((item) => {
          return (
            <button
              key={item.contract || item.symbol}
              className="flex w-full"
              style={{
                height: 36,
                marginTop: 10,
                marginBottom: 10,
              }}
              onClick={() => {
                handleTokenSelect(item);
              }}
              type="button"
            >
              <div style={{ position: 'relative', display: 'inline-block' }}>
                {item.logo ? (
                  <img
                    src={item.logo || ''}
                    style={{
                      width: 36,
                      height: 36,
                      marginLeft: 10,
                      borderRadius: 50,
                    }}
                    alt="token logo"
                  />
                ) : (
                  <div
                    className="w-full h-full overflow-hidden rounded-full"
                    style={{
                      width: 36,
                      height: 36,
                      borderRadius: 50,
                      marginLeft: 10,
                    }}
                  >
                    <RandomAvatar name={item.name || ''} />
                    <span
                      className="absolute inset-0 flex items-center justify-center text-white text-lg font-bold"
                      style={{ marginLeft: 10 }}
                    >
                      {item.name?.slice(0, 2)}
                    </span>
                  </div>
                )}
                <img
                  src={getLogoForChainId(
                    chainNameToChainIdTokensData(item.chain)
                  )}
                  style={{
                    position: 'absolute',
                    bottom: '-2px',
                    right: '-2px',
                    width: 15,
                    height: 15,
                    borderRadius: '50%',
                  }}
                  alt="chain logo"
                />
              </div>
              <div
                className="flex flex-col"
                style={{ width: 250, height: 14, marginLeft: 10 }}
              >
                <div className="flex">
                  <p style={{ fontSize: 13, fontWeight: 400 }}>{item.symbol}</p>
                  <p
                    style={{
                      fontSize: 12,
                      fontWeight: 400,
                      marginLeft: 3,
                      color: 'grey',
                    }}
                  >
                    {item.name}
                  </p>
                </div>
                <div className="flex">
                  <p style={{ fontSize: 13, fontWeight: 400, color: 'grey' }}>
                    MCap:
                  </p>
                  <p
                    style={{
                      fontSize: 13,
                      fontWeight: 400,
                      marginLeft: 3,
                    }}
                  >
                    {formatBigNumber(item.mCap || 0)}
                  </p>
                  <p
                    style={{
                      fontSize: 13,
                      fontWeight: 400,
                      marginLeft: 3,
                      color: 'grey',
                    }}
                  >
                    Vol:
                  </p>
                  <p
                    style={{
                      fontSize: 13,
                      fontWeight: 400,
                      marginLeft: 3,
                    }}
                  >
                    {formatBigNumber(item.volume || 0)}
                  </p>
                </div>
              </div>
              <div className="flex flex-col ml-auto mr-2.5">
                <div>
                  <TokenPrice value={item.price || 0} />
                </div>
                <div className="ml-auto">
                  <TokenPriceChange value={item.priceChange24h || 0} />
                </div>
              </div>
            </button>
          );
        })}
      </>
    );
  }
  // eslint-disable-next-line react/jsx-no-useless-fragment
  return <></>;
}
