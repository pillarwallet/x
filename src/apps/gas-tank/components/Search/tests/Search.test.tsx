/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// types
import { PortfolioData } from '../../../../../types/api';

// hooks
import * as useTokenSearch from '../../../hooks/useTokenSearch';

// utils
import { MobulaChainNames } from '../../../utils/constants';

// components
import Search from '../Search';

// Mock dependencies
vi.mock('../../../hooks/useTokenSearch', () => ({
  useTokenSearch: vi.fn(),
}));

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useLocation: () => ({
      search: '?asset=0x1234567890123456789012345678901234567890',
      pathname: '/',
    }),
    useNavigate: () => vi.fn(),
  };
});

const mockSetSearching = vi.fn();
const mockSetBuyToken = vi.fn();
const mockSetSellToken = vi.fn();
const mockSetChains = vi.fn();

const mockPortfolioData: PortfolioData = {
  assets: [
    {
      asset: {
        id: 1,
        name: 'Test Token',
        symbol: 'TEST',
        logo: 'https://example.com/logo.png',
        decimals: ['18'],
        contracts: ['0x1234567890123456789012345678901234567890'],
        blockchains: ['ethereum'],
      },
      contracts_balances: [
        {
          address: '0x1234567890123456789012345678901234567890',
          balance: 1.0,
          balanceRaw: '1000000000000000000',
          chainId: 'eip155:1',
          decimals: 18,
        },
      ],
      cross_chain_balances: {},
      price_change_24h: 0.05,
      estimated_balance: 1.5,
      price: 1.5,
      token_balance: 1.0,
      allocation: 1.0,
      wallets: ['0x1234567890123456789012345678901234567890'],
    },
  ],
  total_wallet_balance: 1.5,
  wallets: ['0x1234567890123456789012345678901234567890'],
  balances_length: 1,
};

const defaultProps = {
  setSearching: mockSetSearching,
  isBuy: true,
  setBuyToken: mockSetBuyToken,
  setSellToken: mockSetSellToken,
  chains: MobulaChainNames.Ethereum,
  setChains: mockSetChains,
  walletPortfolioData: mockPortfolioData,
  walletPortfolioLoading: false,
  walletPortfolioError: false,
};

const mockUseTokenSearch = {
  searchText: '',
  setSearchText: vi.fn(),
  searchData: null,
  isFetching: false,
};

describe('<Search />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    (useTokenSearch.useTokenSearch as any).mockReturnValue(mockUseTokenSearch);
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Search {...defaultProps} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders main search interface elements', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    expect(screen.getByTestId('pulse-search-view')).toBeInTheDocument();
    expect(screen.getByTestId('pulse-search-modal')).toBeInTheDocument();
    expect(screen.getByTestId('pulse-search-input')).toBeInTheDocument();
    expect(
      screen.getByTestId('pulse-search-filter-buttons')
    ).toBeInTheDocument();
  });

  it('renders buy mode filter buttons', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} isBuy />
      </MemoryRouter>
    );

    expect(screen.getByText('ðŸ”¥ Trending')).toBeInTheDocument();
    expect(screen.getByText('ðŸŒ± Fresh')).toBeInTheDocument();
    expect(screen.getByText('ðŸš€ Top Gainers')).toBeInTheDocument();
    expect(screen.getByText('ðŸ’°My Holdings')).toBeInTheDocument();
  });

  it('renders sell mode with only My Holdings', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} isBuy={false} />
      </MemoryRouter>
    );

    expect(screen.getByText('My Holdings')).toBeInTheDocument();
    expect(screen.queryByText('ðŸ”¥ Trending')).not.toBeInTheDocument();
    expect(screen.queryByText('ðŸŒ± Fresh')).not.toBeInTheDocument();
    expect(screen.queryByText('ðŸš€ Top Gainers')).not.toBeInTheDocument();
  });

  it('handles search input changes', () => {
    const mockSetSearchText = vi.fn();
    (useTokenSearch.useTokenSearch as any).mockReturnValue({
      ...mockUseTokenSearch,
      setSearchText: mockSetSearchText,
    });

    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    const input = screen.getByTestId('pulse-search-input');
    fireEvent.change(input, { target: { value: 'test search' } });

    expect(mockSetSearchText).toHaveBeenCalledWith('test search');
  });

  it('handles filter button clicks in buy mode', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} isBuy />
      </MemoryRouter>
    );

    const trendingButton = screen.getByText('ðŸ”¥ Trending');
    fireEvent.click(trendingButton);

    // Should trigger search type change
    expect(screen.getByText('ðŸ”¥ Trending')).toBeInTheDocument();
  });

  it('shows loading spinner when fetching', () => {
    (useTokenSearch.useTokenSearch as any).mockReturnValue({
      ...mockUseTokenSearch,
      isFetching: true,
      searchText: 'test',
    });

    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    expect(screen.getByTestId('pulse-search-input')).toBeInTheDocument();
  });

  it('shows close button when not fetching', () => {
    (useTokenSearch.useTokenSearch as any).mockReturnValue({
      ...mockUseTokenSearch,
      isFetching: false,
      searchText: 'test',
    });

    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    expect(screen.getByTestId('pulse-search-input')).toBeInTheDocument();
  });

  it('displays My Holdings text when in sell mode', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} isBuy={false} />
      </MemoryRouter>
    );

    expect(screen.getByText('My Holdings')).toBeInTheDocument();
  });

  it('handles token selection for buy mode', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} isBuy />
      </MemoryRouter>
    );

    // Test that the component renders without errors
    expect(screen.getByTestId('pulse-search-view')).toBeInTheDocument();
    expect(screen.getByTestId('pulse-search-modal')).toBeInTheDocument();

    // Test that buy mode shows all filter buttons
    expect(screen.getByText('ðŸ”¥ Trending')).toBeInTheDocument();
    expect(screen.getByText('ðŸŒ± Fresh')).toBeInTheDocument();
    expect(screen.getByText('ðŸš€ Top Gainers')).toBeInTheDocument();
    expect(screen.getByText('ðŸ’°My Holdings')).toBeInTheDocument();
  });

  it('handles token selection for sell mode', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} isBuy={false} />
      </MemoryRouter>
    );

    // Simulate token selection
    const tokenButton = screen.getByText('TEST').closest('button');
    if (tokenButton) {
      fireEvent.click(tokenButton);
    }

    expect(mockSetSellToken).toHaveBeenCalled();
  });

  it('shows search placeholder when no search text and no parsed assets', () => {
    (useTokenSearch.useTokenSearch as any).mockReturnValue({
      ...mockUseTokenSearch,
      searchText: '',
    });

    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    expect(
      screen.getByText('Search by token or paste address...')
    ).toBeInTheDocument();
  });

  it('handles chain overlay toggle', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    const chainButton = screen.getByRole('button', { name: /save/i });
    fireEvent.click(chainButton);

    // Chain overlay should be triggered
    expect(chainButton).toBeInTheDocument();
  });

  it('handles refresh button click', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    const refreshButton = screen.getByRole('button', { name: /refresh/i });
    fireEvent.click(refreshButton);

    expect(refreshButton).toBeInTheDocument();
  });

  it('handles portfolio loading state', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} walletPortfolioLoading />
      </MemoryRouter>
    );

    // Should still render the main search interface
    expect(screen.getByTestId('pulse-search-view')).toBeInTheDocument();
    expect(screen.getByTestId('pulse-search-modal')).toBeInTheDocument();
  });

  it('handles portfolio error state', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} walletPortfolioError />
      </MemoryRouter>
    );

    // Should still render the main search interface
    expect(screen.getByTestId('pulse-search-view')).toBeInTheDocument();
    expect(screen.getByTestId('pulse-search-modal')).toBeInTheDocument();
  });

  it('handles empty portfolio data', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} walletPortfolioData={undefined} />
      </MemoryRouter>
    );

    // Should still render the main search interface
    expect(screen.getByTestId('pulse-search-view')).toBeInTheDocument();
    expect(screen.getByTestId('pulse-search-modal')).toBeInTheDocument();
  });

  it('handles close button click', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    const closeButton = screen.getByRole('button', { name: /close/i });
    fireEvent.click(closeButton);

    expect(mockSetSearching).toHaveBeenCalledWith(false);
  });

  it('auto-focuses search input on mount', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    const input = screen.getByTestId('pulse-search-input');
    expect(input).toBeInTheDocument();
  });

  it('handles URL asset parameter on mount', () => {
    render(
      <MemoryRouter>
        <Search {...defaultProps} />
      </MemoryRouter>
    );

    // Should set search text from URL parameter
    expect(screen.getByTestId('pulse-search-input')).toBeInTheDocument();
  });
});
