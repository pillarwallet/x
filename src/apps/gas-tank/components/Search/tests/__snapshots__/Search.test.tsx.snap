// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Search /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex items-center justify-center min-h-screen bg-black"
  data-testid="pulse-search-view"
>
  <div
    className="flex flex-col w-full max-w-[446px] max-h-[500px] overflow-y-auto bg-[#1E1D24] p-3 border border-white/[0.05] rounded-2xl shadow-[0px_2px_15px_0px_rgba(18,17,22,0.5)]"
    data-testid="pulse-search-modal"
  >
    <div
      className="flex w-full"
    >
      <div
        className="flex items-center justify-center w-3/4 h-10 bg-[#121116] rounded-[10px] m-2.5 border-2 border-[#1E1D24]"
      >
        <span
          className="ml-2.5"
        >
          <img
            alt="search-icon"
            src="data:image/svg+xml,%3csvg%20class='feather%20feather-search'%20fill='none'%20height='24'%20stroke='white'%20stroke-linecap='round'%20stroke-linejoin='round'%20stroke-width='2'%20viewBox='0%200%2024%2024'%20width='24'%20xmlns='http://www.w3.org/2000/svg'%3e%3ccircle%20cx='11'%20cy='11'%20r='8'/%3e%3cline%20x1='21'%20x2='16.65'%20y1='21'%20y2='16.65'/%3e%3c/svg%3e"
          />
        </span>
        <input
          className="flex-1 w-fit ml-4 font-normal text-xs text-gray-500"
          data-testid="pulse-search-input"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          aria-label="Close"
          onClick={[Function]}
          style={
            {
              "height": 20,
              "marginRight": 10,
              "width": 20,
            }
          }
          type="button"
        >
          <img
            alt="close-icon"
            src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20view-box='0%200%2016%2016'%20fill='white'%20class-name='w-5%20h-5'%20%3e%3cpath%20fill-rule='evenodd'%20d='M10%2018a8%208%200%201%200%200-16%208%208%200%200%200%200%2016ZM8.28%207.22a.75.75%200%200%200-1.06%201.06L8.94%2010l-1.72%201.72a.75.75%200%201%200%201.06%201.06L10%2011.06l1.72%201.72a.75.75%200%201%200%201.06-1.06L11.06%2010l1.72-1.72a.75.75%200%200%200-1.06-1.06L10%208.94%208.28%207.22Z'%20clip-rule='evenodd'%20/%3e%3c/svg%3e"
          />
        </button>
      </div>
      <div
        className="mt-2.5 w-10 h-10 bg-black rounded-[10px] p-0.5 pb-1 pl-0.5 pr-0.5"
      >
        <button
          aria-label="Refresh"
          className="flex items-center justify-center w-[36px] h-[34px] bg-[#1E1D24] rounded-[8px] cursor-not-allowed opacity-50"
          disabled={true}
          type="button"
        >
          <img
            alt="refresh-icon"
            src="data:image/svg+xml,%3csvg%20width='20'%20height='18'%20viewBox='0%200%2020%2018'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cg%20opacity='0.5'%3e%3cpath%20d='M1.93781%209.03579C1.94148%207.2726%202.52824%205.55746%203.6071%204.1564C4.68595%202.75533%206.19659%201.74663%207.90475%201.28674C9.6129%200.826854%2011.4231%200.941474%2013.0546%201.61283C14.6861%202.28419%2016.0477%203.47475%2016.9282%204.99989M16.9282%204.99989L13.7631%204.51784M16.9282%204.99989L16.7272%201.65181'%20stroke='white'%20stroke-width='2'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20opacity='0.5'%20d='M18.0619%208.96372C18.0583%2010.7269%2017.4715%2012.442%2016.3927%2013.8431C15.3138%2015.2442%2013.8032%2016.2529%2012.095%2016.7128C10.3869%2017.1727%208.57666%2017.058%206.94518%2016.3867C5.31369%2015.7153%203.9521%2014.5248%203.07156%2012.9996M3.07156%2012.9996L6.23662%2013.4817M3.07156%2012.9996L3.27252%2016.3477'%20stroke='white'%20stroke-width='2'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/g%3e%3c/svg%3e"
          />
        </button>
      </div>
      <div
        className="ml-1.5 mt-2.5 w-10 h-10 bg-black rounded-[10px] p-0.5 pb-1 pl-0.5 pr-0.5 relative cursor-pointer"
        onClick={[Function]}
      >
        <button
          aria-label="Save"
          className="flex items-center justify-center w-full h-full bg-[#121116] rounded-[10px]"
          type="button"
        >
          <img
            alt="globe-icon"
            src="data:image/svg+xml,%3csvg%20width='20'%20height='20'%20viewBox='0%200%2020%2020'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cg%20opacity='0.5'%3e%3ccircle%20opacity='0.3'%20cx='9.99976'%20cy='10'%20r='9'%20fill='white'/%3e%3cpath%20d='M9.99992%2019.0002C14.9706%2019.0002%2019.0001%2014.9706%2019.0001%2010C19.0001%205.02934%2014.9706%200.999832%209.99992%200.999832C5.02927%200.999832%200.999756%205.02934%200.999756%2010C0.999756%2014.9706%205.02927%2019.0002%209.99992%2019.0002Z'%20stroke='white'%20stroke-width='1.5'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20d='M6.3999%201.89984H7.29992C5.54488%207.15594%205.54488%2012.844%207.29992%2018.1001H6.3999'%20stroke='white'%20stroke-width='1.5'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20d='M12.7%201.89984C14.455%207.15594%2014.455%2012.844%2012.7%2018.1001'%20stroke='white'%20stroke-width='1.5'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20d='M1.90015%2013.6001V12.7C7.15624%2014.4551%2012.8443%2014.4551%2018.1004%2012.7V13.6001'%20stroke='white'%20stroke-width='1.5'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20d='M1.90015%207.29995C7.15624%205.54492%2012.8443%205.54492%2018.1004%207.29995'%20stroke='white'%20stroke-width='1.5'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/g%3e%3c/svg%3e"
          />
        </button>
      </div>
    </div>
    <div
      className="flex gap-1.5"
      data-testid="pulse-search-filter-buttons"
    >
      <div
        className="flex bg-black w-[100px] h-10 rounded-[10px]"
      >
        <button
          className="flex-1 items-center justify-center rounded-[6px] m-0.5 mb-1 bg-[#1E1D24]"
          onClick={[Function]}
          type="button"
        >
          <p
            className="text-xs font-normal text-center text-white opacity-50"
          >
            ðŸ”¥ Trending
          </p>
        </button>
      </div>
      <div
        className="flex bg-black w-[100px] h-10 rounded-[10px]"
      >
        <button
          className="flex-1 items-center justify-center rounded-[6px] m-0.5 mb-1 bg-[#1E1D24]"
          onClick={[Function]}
          type="button"
        >
          <p
            className="text-xs font-normal text-center text-white opacity-50"
          >
            ðŸŒ± Fresh
          </p>
        </button>
      </div>
      <div
        className="flex bg-black w-[100px] h-10 rounded-[10px]"
      >
        <button
          className="flex-1 items-center justify-center rounded-[6px] m-0.5 mb-1 bg-[#1E1D24]"
          onClick={[Function]}
          type="button"
        >
          <p
            className="text-xs font-normal text-center text-white opacity-50"
          >
            ðŸš€ Top Gainers
          </p>
        </button>
      </div>
      <div
        className="flex bg-black w-[100px] h-10 rounded-[10px]"
      >
        <button
          className="flex-1 items-center justify-center rounded-[6px] m-0.5 mb-1 bg-[#1E1D24]"
          onClick={[Function]}
          type="button"
        >
          <p
            className="text-xs font-normal text-center text-white opacity-50"
          >
            ðŸ’°My Holdings
          </p>
        </button>
      </div>
    </div>
    <div
      className="flex items-center justify-center m-[50px]"
    >
      <p
        className="text-gray-500"
      >
        Search by token or paste address...
      </p>
    </div>
  </div>
</div>
`;
